JSON & YAML tests

Pluies opened the pull request:
Early prototype, but opening a PR to start the discussion on the approach and get some feedback :)

In order to support JSON configuration (fixes #195) which is the default in EKS images (would therefore also help #220), I propose a new operator for the `test` structure, `jsonpath`. This uses Kubernetes' [jsonpath helper](https://kubernetes.io/docs/reference/kubectl/jsonpath/) to look up configuration values in a JSON file rather than as flags in a configuration file.

Besides the audit source and the `flag` -> `jsonpath` change, the check stays identical. For example, the existing check:
```yaml
  - id: 2.1.2
    text: "Ensure that the --anonymous-auth argument is set to false (Scored)"
    audit: "ps -fC $kubeletbin"
    tests:
      test_items:
      - flag: "--anonymous-auth"
        compare:
          op: eq
          value: false
        set: true
```
Can be written as:
```yaml
  - id: 2.1.2
    text: "Ensure that the --anonymous-auth argument is set to false (Scored)"
    audit: "cat $kubeletconf"
    tests:
      test_items:
      - jsonpath: "{.authentication.anonymous.enabled}"
        compare:
          op: eq
          value: "false"
        set: true
```
And will behave as the `flag` operator viz paramaters `set` and `compare`, i.e.:
- If `.authentication`, `.authentication.anonymous`, or `.authentication.anonymous.enabled` are unset: fail
- If `.authentication.anonymous.enabled` is not set to "false": fail
- If `.authentication.anonymous.enabled` is set to "false": true

I've written this for EKS for now, adding a `1.11-eks` profile, but it's 100% compatible with other implementations that also use json config files for the kubelet. This approach is also backwards-compatible with existing checks that use the `flag` param.

My end goal with this approach is to eventually get rid of the need to reading the configuration file locally, and instead [directly retrieve the current kubelet configuration in JSON format through the Kubernetes API](https://kubernetes.io/docs/tasks/administer-cluster/reconfigure-kubelet/#generate-the-configuration-file). This offers two main benefits:
- Default values will be filled in by the API, making for easier testing
- This approach supports [dynamic kubelet configuration](https://kubernetes.io/blog/2018/07/11/dynamic-kubelet-configuration/), where the configuration on a running node may have been updated compared to its on-disk config

Is this something that I should continue forward with (write tests, make more resilient, update docs), or is there already a different plan to achieve this?

cc @lizrice

alexbrand commented:
I was going to start experimenting on an approach for this. One thing that might be a challenge is that AFAIK configuration can be specified in both flags and the config file.

alexbrand commented on cfg/1.11-eks/node.yaml:
```diff
@@ -0,0 +1,526 @@
+    scored: true
+
+  - id: 2.1.11
+    text: "Ensure that the --tls-cert-file and --tls-private-key-file arguments are set as appropriate (Scored)"
```
This is probably out of scope for this PR, but I wonder what is the reasoning for this recommendation, as the kubelet is automatically requesting and rotating kubelet serving certs (due to `RotateKubeletServerCertificate` being set to `true`)

lizrice replied:
This is really a question for the CIS as we try not to diverge from what's in the benchmark

Pluies replied:
I noticed this as well, and I think @lizrice is right, this should be fixed at the benchmark level.

lizrice commented:
I really like the general approach for this, thanks @Pluies!

Agree with @alexbrand that the settings in a config file can be overridden by command line parameters. We could start with essentially what's here (with the config in a separate directory that has to be explicitly specified with `--version`) - that would be useful for EKS installations today. A next step would combine the two approaches so we can have checks for both the command line parameter and the config file setting within one test file. With that in mind, maybe rather than changing the audit command to always `cat` a config file, we could define another field (say, `configfile`) so that both the audit command and the config file can be defined within the same test. Wdyt?

Two other thoughts: 
* AFAIK config files could also be specified in YAML - is there an easy way to have this approach support both formats? 

* I'd suggest we call the directory 1.11-configfile (or similar) because it could be used on other installations that use the config file approach, not just EKS. 

@ttousai any additional thoughts about this approach?

ttousai commented:
I think this approach is great. But I have pretty much the same questions as @lizrice.

ttousai commented:
@lizrice @Pluies I have an idea how to solve some our questions on this PR. Keep in mind kube-bench was designed original to run only commands and check for flags, but now we need to support checking values in config file too.

So how about we add type of audit as subgroups under audit and chain audit types. There can be a command audit type and a file audit type. These subgroups will be run in the order they are defined (chained) and the first to match successfully will terminate the audit chain.

```
...
audit:
  command:
    cmd: "some command --flag1=value1"
    tests:
      test_items:
        flag: "--flag"
        compare:
          op: eq
          value: value1
        set: true

  file:
    path: /path/to/configfile
    tests:
       test_items:
         jsonpath: "{.flag}"
         compare:
           op: eq
           value: value1
         set: true
  
  file:
    path: /path/to/configfile
    tests:
       test_items:
         yamlpath: "flag"
         compare:
           op: eq
           value: value1
         set: true
....
```
In the example above if the flag is present in the command line parameters, the command audit type will run and terminate the audit chain, else the file audit type will run with jsonspath, then jsonpath if all else fails.

lizrice commented:
The concern I have with this YAML layout is that we need to repeat the test part: in all the cases the bit that varies is how we get the flag and its value, but the comparison (e.g. == 0) is always the same (and I think always will be, otherwise it would be a different test. 

But we could avoid the repetition with something like: 

audit:
  command: 
  file: 
  tests: 

That is, have command, file and tests fields all at the same level. Wdyt? 

(sorry for bad formatting, can't find back-tick on my phone!)

ttousai commented:
@lizrice the repetition can be a pain. We could have shared tests which will be run for all audits types (much like the current structure) and audit type specify tests in the event that different tests must be run for the different audit types.

But to get started we could implement that shared tests and then implement the separate test by audit type if there is a need.

alexbrand commented:
When I first started to think about this it seemed like it might make sense to split the process into multiple steps: 1) obtain the configuration and support multiple sources (flags, config file, remote API call), 2) create a data structure that holds the configuration, and then 3) run tests agains the data structure. 

Not sure if this helps though, as there are some tests that do not run against the component configuration (e.g. permissions on files).

Pluies commented:
> When I first started to think about this it seemed like it might make sense to split the process into multiple steps: 1) obtain the configuration and support multiple sources (flags, config file, remote API call), 2) create a data structure that holds the configuration, and then 3) run tests agains the data structure.

That's fair, it's probably the best way forward. And to be fair, the API will get us most of (if not all) the way there as it already gathers configuration from flags / config files / configmap updates in case of dynamic reconfiguring, then serialises it into the "official" k8s data structure that we can query: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/config/types.go

This may take a little effort to get in a nice place, so I'll start with updating this PR to get a first go as outlined by Liz:
> We could start with essentially what's here (with the config in a separate directory that has to be explicitly specified with --version) - that would be useful for EKS installations today.

Hopefully should have something merge-able this week :)

christopherhein commented:
/cc @mhausenblas

Pluies commented on cfg/1.11-eks/node.yaml:
```diff
@@ -0,0 +1,526 @@
+        compare:
+          op: noteq
+          value: "0"
+        set: true
```
Fixed this 👍

Pluies commented on cfg/1.11-eks/node.yaml:
```diff
@@ -0,0 +1,526 @@
+    scored: true
+
+  - id: 2.1.8
+    text: "Ensure that the --make-iptables-util-chains argument is set to true (Scored)"
```
Adding this too 👍

Pluies commented:
This is now ready for reviewing :)

I've added support for YAML, which turned out to be super easy: both YAML and JSON decode in an `interface{}` type that I can then use just like I used the json. 👍 

I've added tests for the new functionalities, which bring `test.go` to 88.2% test coverage, and the project overall to 53.1%.

As for the actual goal, which was to support EKS better, here are the results on the default EKS AMI with the existing `1.11` profile:

<img width="947" alt="1 11" src="https://user-images.githubusercontent.com/344835/54750913-3458e580-4bd1-11e9-9a93-f00e06bd8a48.png">

And with the new `1.11-json` profile:

<img width="948" alt="1 11-json" src="https://user-images.githubusercontent.com/344835/54750915-36bb3f80-4bd1-11e9-955d-5f3668877325.png">

As far as I can tell, the remaining issues are genuine issues where EKS differs from CIS:
- 2.1.1: `allowPrivileged` is needed by Amazon's VPC CNI Plugin
- 2.1.7: `protectKernelDefaults` is unset
- 2.1.10: event qps is unset and defaults to a value of 5
- 2.1.11: `--tls-cert-file` and `--tls-private-key-file` are not needed in EKS as it uses IAM + instance profiles to establish connections with the apiserver, but as mentioned in a comment above this is something that needs to be changed in the benchmark

alexbrand commented on cfg/1.11-json/node.yaml:
```diff
@@ -0,0 +1,508 @@
+  checks:
+  - id: 2.1.1
+    text: "Ensure that the --allow-privileged argument is set to false (Scored)"
+    audit: "ps -fC $kubeletbin"
```
Should this be checking the config file instead of flags?

Pluies replied:
Well... I thought so at first, but then couldn't find anything related to allow-privileged in the configuration format. According to [the docs](https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/):

> The subset of the Kubelet’s configuration that can be configured via a file is defined by the `KubeletConfiguration` struct [here (v1beta1)](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/config/types.go).

And the struct it points to does not have a member for this option, so I assume this configuration is not part of the "subset" that can be defined in JSON or through dynamic configuration and therefore _has_ to be passed at kubelet start-up as an argument.

alexbrand replied:
Huh! I totally assumed you could, but I guess not 😄

Pluies commented on README.md:
````diff
@@ -159,6 +159,29 @@ tests:
 ...
 ```
+
+If using a JSON config file, the syntax is:
````
Good points re this and the next comments. What do you think would make more sense? Something like this, listing all three types of tests?

> The syntax for tests:
> 
> ```
> tests:
> - flag:
>   set:
>   compare:
>     op:
>     value:
> - jsonpath:
>   set:
>   compare:
>     op:
>     value:
> - yamlpath:
>   set:
>   compare:
>     op:
>     value:
> ```

alexbrand replied:
I kinda like that you've separated them into separate code blocks. Maybe the sentence could read something like:

```
You can also define jsonpath and yamlpath tests using the following syntax:
```

Pluies replied:
Thanks, that's nice and clear :) Updated the readme.

Pluies commented on README.md:
```diff
@@ -149,7 +149,7 @@ These groups are further organized under `controls` which can be of the type `ma
 Tests are the items we actually look for to determine if a check is successful or not. Checks can have multiple tests, which must all be successful for the check to pass.
 
-The syntax for tests:
+The syntax for tests operating on a flag:
```
👍 fixed

Pluies commented on README.md:
````diff
@@ -159,6 +159,29 @@ tests:
+...
+```
+
+And for a YAML config file:
````
👍 fixed

djsly commented:
@Pluies I am setup to test this if you need. I got most of my config for kubelet stored in yaml format.

Pluies commented:
@djsly sounds good! I've pushed the docker image at [pluies/kube-bench:json-config](https://hub.docker.com/r/pluies/kube-bench/tags), so feel free to run it with the `--version 1.11-json` argument to test over JSON files. You can also pull this branch and build your own container if easier or if you need to tweak the config 👍 

@lizrice @alexbrand , anything I can do to help getting this through?

lizrice commented:
I couldn't actually figure out how to push more changes to this feature branch so I created a new PR, #260, that builds on @Pluies great work and just tidies up a few things. Would be great if anyone has the time to try that one out and confirm if it does the trick for you? cc @djsly @alexbrand  @mhausenblas

lizrice commented:
I think I've incorporated all this under #260 so I'm going to close this one now. Please shout if you have any concerns. Thank you so much everyone, especially @Pluies, for all your work on this!

Pluies commented:
Oh, that's brilliant, thanks for building upon it and getting this merged @lizrice ! :)
